src/python/pants/jvm/resolve/jvm_tool.py

```diff
@@ -0,0 +1,334 @@
+
+    # Default resource for the tool's lockfile. (Subclasses must set.)
+    default_lockfile_resource: ClassVar[tuple[str, str]]
+
+    default_lockfile_url: ClassVar[str | None] = None
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--version",
+            type=str,
+            advanced=True,
+            default=cls.default_version,
+            help=(
+                "Version string for the tool. This is available for substitution in the "
+                f"`[{cls.options_scope}].artifacts` option by including the string "
+                "`{version}`."
+            ),
+        )
+        register(
+            "--artifacts",
+            type=list,
+            member_type=str,
+            advanced=True,
+            default=cls.default_artifacts,
+            help=(
+                "Artifact requirements for this tool using specified as either the address of a `jvm_artifact` "
+                "target or, alternatively, as a colon-separated Maven coordinates (e.g., group:name:version). "
+                "For Maven coordinates, the string `{version}` version will be substituted with the value of the "
+                f"`[{cls.options_scope}].version` option."
+            ),
+        )
+        register(
+            "--lockfile",
+            type=str,
+            default=cls.default_lockfile_path,
+            advanced=True,
+            help=(
+                "Path to a lockfile used for installing the tool.\n\n"
+                f"Set to the string `{DEFAULT_TOOL_LOCKFILE}` to use a lockfile provided by "
+                "Pants, so long as you have not changed the `--version` option. "
+                f"See {cls.default_lockfile_url} for the default lockfile contents.\n\n"
+                "To use a custom lockfile, set this option to a file path relative to the "
+                f"build root, then run `./pants jvm-generate-lockfiles "
+                f"--resolve={cls.options_scope}`.\n\n"
+            ),
+        )
+
+    @property
+    def version(self) -> str:
+        return cast(str, self.options.version)
+
+    @property
+    def artifact_inputs(self) -> tuple[str, ...]:
+        return tuple(s.format(version=self.version) for s in self.options.artifacts)
+
+    @property
+    def lockfile(self) -> str:
+        f"""The path to a lockfile or special string '{DEFAULT_TOOL_LOCKFILE}'."""
+        lockfile = cast(str, self.options.lockfile)
+        if lockfile != DEFAULT_TOOL_LOCKFILE:
+            return lockfile
+        pkg, path = self.default_lockfile_resource
+        return f"src/python/{pkg.replace('.', '/')}/{path}"
+
+    def lockfile_content(self) -> bytes:
+        lockfile_path = self.lockfile
+        if lockfile_path == DEFAULT_TOOL_LOCKFILE:
+            return importlib.resources.read_binary(*self.default_lockfile_resource)
+        with open(lockfile_path, "rb") as f:
+            return f.read()
```

Comment:
This doesn't seem safe with caching?

Reply:
I didn't really consider the caching aspect. Another thing to fix ...

Reply:
I'm addressing this as part of #13827.